tensorflow_datasets/image_classification/image_folder.py

```diff
@@ -29,147 +29,72 @@
+    self._name = dataset_name
+    self._data_dir = data_dir or os.path.expanduser(
+      os.path.join(DATA_DIR, dataset_name))
+
+    # dict[split_name][label_name] = list(img_paths)
+    self._split_label_images = self._get_split_label_images()
+
+  def _get_split_label_images(self):
+    split_names = list_folders(self.data_dir)
+    split_label_images = {}
     for split_name in split_names:
-      split_dir = os.path.join(dl_manager.manual_dir, split_name)
+      split_dir = os.path.join(self.data_dir, split_name)
       split_label_images[split_name] = {
-          label_name: list_imgs(os.path.join(split_dir, label_name))
-          for label_name in list_folders(split_dir)
+          label_name: list_imgs(os.path.join(split_dir, label_name))	
+          for label_name in list_folders(split_dir)	
+      }
+    return split_label_images
+
+  @property
+  def data_dir(self):
+    return self._data_dir
+
+  @property
+  def name(self):
+    return self._name
+
+  def as_dataset(self, split=None):
+    if split:
+      return self._as_dataset(split)
+    return {
+          split_name: self._as_dataset(split_name)
+          for split_name in list_folders(self.data_dir)
       }
 
-    # Merge all label names from all splits to get the final list of labels
-    # Sorted list for determinism
-    labels = [split.keys() for split in split_label_images.values()]
-    labels = list(sorted(set(itertools.chain(*labels))))
-
-    # Could improve the automated encoding format detection
-    # Extract the list of all image paths
-    image_paths = [
-        image_paths
-        for label_images in split_label_images.values()
-        for image_paths in label_images.values()
-    ]
-    if any(f.lower().endswith(".png") for f in itertools.chain(*image_paths)):
-      encoding_format = "png"
-    else:
-      encoding_format = "jpeg"
-
-    # Update the info.features. Those info will be automatically resored when
-    # the dataset is re-created
-    self.info.features["image"].set_encoding_format(encoding_format)
-    self.info.features["label"].names = labels
-
-    # Define the splits
-    return [
-        tfds.core.SplitGenerator(
-            name=split_name,
-            gen_kwargs=dict(label_images=label_images,),
-        ) for split_name, label_images in split_label_images.items()
-    ]
-
-  def _generate_examples(self, label_images):
-    """Generate example for each image in the dict."""
-
-    for label, image_paths in label_images.items():
-      for image_path in image_paths:
-        key = "%s/%s" % (label, os.path.basename(image_path))
-        yield key, {
-            "image": image_path,
-            "label": label,
-        }
+  def _as_dataset(self, split_name):
+    if split_name not in self._split_label_images.keys():
+      raise ValueError("Split name {} not present in {}".format(split_name, self._split_label_images.keys()))
 
+    imgs = [img for l in self._split_label_images[split_name].values() for img in l]
```

Comment:
The list of images should be shuffled deterministically. (So that two users have examples in the same order).

When `shuffle_files=True`, there should be an additional `ds = ds.shuffle(len(image_paths))` after the `tf.data.Dataset.list_files`.